proxy/src/control/cache.rs

```diff
@@ -37,15 +41,20 @@ impl<T> Exists<T> {
 
 // ===== impl Cache =====
 
-impl<T> Cache<T> where T: Clone + Copy + Eq + std::hash::Hash {
+impl<K, V> Cache<K, V>
+where
+    K: Copy + Clone,
+    K: Hash + Eq,
+    V: Clone,
+{
     pub fn new() -> Self {
         Cache {
-            values: HashSet::new(),
+            values: IndexMap::new(),
             reset_on_next_modification: true,
         }
     }
 
-    pub fn values(&self) -> &HashSet<T> {
+    pub fn values(&self) -> &IndexMap<K, V> {
```

Comment:
nit: perhaps better to rename `values` to  something else, as it makes one think it would return an iterator of `V`s